### ISSUE TITLE
Question: multi-scale loss weighting

### ISSUE THREAD

**Issue by jbohnslav** [ClementPinard/FlowNetPytorch#54]
Hello,

In the [Nvidia implementation of FlowNet2](https://github.com/NVIDIA/flownet2-pytorch/blob/master/losses.py), they compute default weights for the multi-scale loss in the following manner: 
```
self.loss_weights = torch.FloatTensor([(l_weight / 2 ** scale) for scale in range(self.numScales)])
```
This outputs `tensor([0.3200, 0.1600, 0.0800, 0.0400, 0.0200])`. This means that **the highest-resolution flow is weighted most highly.** In your implementation, the default loss weights are  
```
weights = [0.005, 0.01, 0.02, 0.08, 0.32]  # as in original article
``` 
This will weight the lowest resolution most highly. 

In the original paper by Fischer, Dosovitskiy, et al., they note that "as a training loss we use the endpoint error... It is the Euclidian distance between the predicted flow vector and the ground truth, averaged over all pixels." [Paper here.](https://arxiv.org/abs/1504.06852). It makes no mention of multi-scale loss functions, or how they are weighted.

**TL;DR: How should you weight the multi-scale loss, and from what research paper is this loss from?**

**Comment by ClementPinard** [comment]
The paper does not mention multi scale loss, but the caffe code they provide does ! It can be read in the model's prototxt in caffe : look for `flow_lossN` elements in train.prototxt, N varying from 2 to 6.

So why is it different from flownet2 pytorch architecture ? Because we compute the sum of EPE, and not mean. 
compare [here](https://github.com/NVIDIA/flownet2-pytorch/blob/master/losses.py#L12) and [here](https://github.com/ClementPinard/FlowNetPytorch/blob/master/multiscaleloss.py#L16)

When mean is used, the "per-pixel" loss (and thus gradient) is much higher for low scale flow maps because it gets divided by the number of pixels. Hence the lower weight when using mean EPE in NVIDIA implementation.

It was actually a problem when I first implemented the network because I was using mean EPE, and multi scale caffe loss, which are adapted for sum EPE. See here for a topic about it : #17

**Comment by jbohnslav** [comment]
This was very helpful, thank you! The difference between summing and averaging the EPE-maps explains the different weights.

I find it very frustrating that the authors did not fully describe something as important as their loss function in the paper, but this is not your fault! Closing the issue.
